19

ehmkah opened the pull request:
This is my first draft for configuring different publish and reply message brokers.

I am open for your comments. Especially for the simple test-setup I provied (which can be deleted if you are not interested in it)

Tomorrow I am going to test it with our system.

ehmkah commented:
I tested it with our system and it works. @red-bashmak - would you be so kind and have a look?

ehmkah commented:
Hi @jigarkhwar - maybe you can have a look at this PR? Is there a simplier way to contact one of you than those PRs?

jigarkhwar commented:
> Hi @jigarkhwar - maybe you can have a look at this PR? Is there a simplier way to contact one of you than those PRs?

Hi, Michael, We have a Christmas and New year holidays till to 11 January. I can check this PR after holidays, sorry!

red-bashmak commented:
@ehmkah thanks for your contribution. But, I have some doubts about using 2 pools to get responses and send requests. It seems to me that there will be problems when some more intermediate brokers need to be added. Then for intermediate brokers will you need to add more storage? In my opinion, some more universal solution is needed.

ehmkah commented:
Hi @red-bashmak - than you for your reply. I am not sure whether I understood your point.  Is it possible that you describe the example you are thinking of? Because I can not see a case where intermediate brokers may be useful. The reason I dont see it is not that I think its not useful its more that I only have an good understanding of our system and the features we need.

If I understand your point I can bake it in.

Just to have an idea what I have in mind, when I made the PR:
- All messages are send to the request broker and all message are taken from the response broker. It simplifies writing scenarios because its easy to check where the messages arrive.

red-bashmak commented on src/test/scala/ru/tinkoff/gatling/amqp/examples/RequestReplyGatlingRunner.scala:
```diff
@@ -0,0 +1,27 @@
+ * allows us to do some configuration and setup before
+ * Gatling launches.
+ */
+object RequestReplyTwoBrokerRunnerRunner {
```
Typo in object name - 2 time Runner

ehmkah replied:
fixed.

red-bashmak commented on src/test/scala/ru/tinkoff/gatling/amqp/examples/RabbitMQConsumer.scala:
```diff
@@ -0,0 +1,79 @@
+  private val writePort = 5673
+
+  def tearDown(): Unit = {
+    val readConnection: Connection = getConnection(readPort)
```
Better make global readConnection and writeConnection in upper of object. Analogously for channels except maybe channel in DeliverCallback, because this channel in in consumer thread

ehmkah replied:
i made the readChannel and writechannel to global variables. The readConnection was always used in conjunction with the readChannel (same for writeChannel) and so its not need. I am brave enough to say, this has been fixed.

red-bashmak commented on src/test/scala/ru/tinkoff/gatling/amqp/examples/RabbitMQConsumer.scala:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  def readAndWrite() = {
+    val connection = getConnection(readPort)
```
See comment bellow

ehmkah replied:
fixed

red-bashmak commented on src/test/scala/ru/tinkoff/gatling/amqp/examples/RabbitMQConsumer.scala:
```diff
@@ -0,0 +1,79 @@
+
+  def setup(): Unit = {
+    val readConnection: Connection = getConnection(readPort)
+    val readChannel = readConnection.createChannel()
```
See comment about global connections

ehmkah replied:
fixed

red-bashmak reviewed (changes requested):
Need refactor RabbitMQConsumer

red-bashmak commented:
@ehmkah Hello, Michael! Okay, maybe problems during expansion will not occur. Perhaps intermediate brokers are not needed. I will accept the merge request, but please transfer the RabbitMQConsumer object to Utils. And it would be better to run RabbitMQConsumer logic in a separate thread along with the scenario in Runner. Or use some kind of client for rabbit and add it to docker compose. And i think need some refactoring in RabbitMQConsumer, i described it in comments

ehmkah commented:
@red-bashmak - Thank you for answer. I am going to fix, your findings and write you a message if I am done.

ehmkah commented on src/test/scala/ru/tinkoff/gatling/amqp/examples/RabbitMQConsumer.scala:
```diff
@@ -0,0 +1,79 @@
+
+// Testclient which consumes messages and writes something in other queue
+
+object RabbitMQConsumer {
```
I moved it to ru...example.utils and renamed it to `SimpleRabbitMQClient`. Its not used in code directly its just needed if someone wants to try out the two broker example. It remains in the test directory of course.

ehmkah commented:
Hi @red-bashmak - I hope I applied all fixed in a way which is good for you. 
- the client is started from the example test in `before`and at the end of the tests is deleted in `after`
- I updated the given documentation to relflect the changes
- I added directory `results`to .gitignore because after some test-runs  I was a little bit afraid to commit them

Happy to hear something from you

ehmkah commented:
Hi @red-bashmak @jigarkhwar - Some days passed, had you already time to review my changes? To give you some extra motivation: If this PR gets merged I am willing to upgrade this plugin to new gatling and scala version (issue number 23) :-)

ehmkah commented:
Thank you a lot. Do you know when you will publish a new version of the plugin?
